### PR TITLE
WT-13124 Give WT_INTL_INDEX_GET acquire semantics, add TSan support

### PR DESCRIPTION
**Pull request by ajmorton**
Update `WT_INTL_INDEX_GET_SAFE` to add an `ACQUIRE_BARRIER` that pairs with the `RELEASE_BARRIER` already present in `WT_INTL_INDEX_SET`. 

This change also adds a TSan compatible implementation of the `WT_INTL_INDEX_*` macros which use __atomics instead of barriers. The semantics of both implementations are identical, but __atomics can be understood by TSan.

### CODE REVIEW

File: test/evergreen/tsan_warnings.supp [hunk 1]
```diff
@@ -29,7 +29,10 @@ race:__ref_track_state
 race:__wt_configure_method
 
 # FIXME-WT-13074 This Work has been split out into a separate ticket to keep PR sizes down.
-race:__wt_calloc
+race:__wt_free_int
```
Comment (ajmorton) [inline test/evergreen/tsan_warnings.supp:32]: The warnings were revealed by the `WT_INTL_INDEX_*` fix. I'll address them in WT-13074

**Comment by ajmorton** [comment]
Perf testing shows no changes. I've attached the (developer only) results to the Jira ticket.

File: src/btree/bt_split.c [hunk 2]
```diff
@@ -510,7 +513,10 @@ __split_root(WT_SESSION_IMPL *session, WT_PAGE *root)
      * Confirm the root page's index hasn't moved, then update it, which makes the split visible to
      * threads descending the tree.
      */
-    WT_ASSERT(session, WT_INTL_INDEX_GET_SAFE(root) == pindex);
+#ifdef HAVE_DIAGNOSTIC
+    WT_INTL_INDEX_GET_SAFE(root, check_root);
```
Comment (luke-pearson) [inline src/btree/bt_split.c:517]: Can you elaborate on this?

File: src/include/btmem.h [hunk 3]
```diff
@@ -616,17 +616,41 @@ struct __wt_page {
...
+    } while (0)
+#define WT_INTL_INDEX_SET(page, v)                                        \
+    do {                                                                  \
+        __atomic_store_n(&(page)->u.intl.__index, (v), __ATOMIC_RELEASE); \
+    } while (0)
+#else
+#define WT_INTL_INDEX_GET_SAFE(page, pindex)                          \
+    do {                                                              \
+        WT_ACQUIRE_BARRIER();                                         \
+        (pindex) = __wt_atomic_load_pointer(&(page)->u.intl.__index); \
```
Comment (luke-pearson) [inline src/include/btmem.h:641]: This should be a WT_ACQUIRE_READ().

File: src/btree/bt_split.c (hunk 2 above)
Comment (luke-pearson) [reply src/btree/bt_split.c:517]: Sorry misread code.

File: src/include/btmem.h (hunk 3 above)
Comment (ajmorton) [reply src/include/btmem.h]: Updated in [399fa38](https://github.com/wiredtiger/wiredtiger/pull/10707/commits/399fa3821e03e50e02fe9657412386c6f58d4f74). I'd actually put the barrier on the wrong side of the instruction in the original commit.

One thing to call out here is `WT_INTL_INDEX_GET_SAFE` will now use `ldapr` instructions on compatible platforms. We're getting acquire semantics so I assume we maintain the atomicity guarantees of the original `__wt_atomic_load_pointer`.

Comment (luke-pearson) [reply src/include/btmem.h]: > One thing to call out here is WT_INTL_INDEX_GET_SAFE will now use ldapr instructions on compatible platforms

Should give the guarantees we want. We're going from a raw load from memory to a RCpc load.

File: src/include/btmem.h [hunk 4]
```diff
@@ -616,17 +616,37 @@ struct __wt_page {
...
         WT_ASSERT(session, __wt_session_gen(session, WT_GEN_SPLIT) != 0); \
-        (pindex) = WT_INTL_INDEX_GET_SAFE(page);                          \
+        WT_INTL_INDEX_GET_SAFE(page, (pindex));                           \
+    } while (0)
+#define WT_INTL_INDEX_SET(page, v)                                        \
+    do {                                                                  \
+        __atomic_store_n(&(page)->u.intl.__index, (v), __ATOMIC_RELEASE); \
+    } while (0)
+#else
+#define WT_INTL_INDEX_GET_SAFE(page, pindex) WT_ACQUIRE_READ((pindex), (page)->u.intl.__index)
```
[the remark this answers is not in the document]
Comment (ajmorton) [reply src/include/btmem.h:639]: [dc90c79](https://github.com/wiredtiger/wiredtiger/pull/10707/commits/dc90c7900a0be75484d266d0df572f1d4899311a)

**Review by luke-pearson — APPROVED** [review]
LGTM.

**Pull request merged**